robot_smach_states/tutorial/challenge_tutorial.md

```diff
@@ -0,0 +1,236 @@
+
+There are various types of designators defined in `robot_skills.util.designators` for various purposes. In practice, however, custom designators are often defined to meet challenge-specific needs.
+
+## Exercise 2
+* Add the following Smach to your `example.py` (fix the import whenever necessary) that will choose a random furniture object in your environment to navigate to:
+```python
+class GenerateRandomTarget(smach.State):
+    """ Smach state that puts a random target into the provided designator
+
+    """
+    def __init__(self, robot, goal_designator):
+        """ Initialization method
+
+        :param robot: (Robot) robot api object
+        :param goal_designator:
+        """
+        super(GenerateRandomTarget, self).__init__(outcomes=["done"])
+        self._robot = robot
+        self._writeable_designator = states.util.designators.core.VariableWriter(goal_designator)
+
+    def execute(self, ud):
+        """ Executes the state. Chooses a random target and puts this in the designator
+
+        :param ud: {}
+        :return: "done"
+        """
+        # Get all entities
+        entities = self._robot.ed.get_entities()
+
+        # Filter the entity id's (we're only interested in furniture objects) and choose one
+        target_entity = random.choice([e for e in entities if e.is_a("furniture")])
+        self._robot.speech.speak("Someone told me I should go to the {}".format(target_entity.id))
+
+        # Set the current id
+        self._writeable_designator.write(target_entity)
+
+        return "done"
+```
+As is described in the comments, the line `self._writeable_designator.write(target_entity)` makes sure that the entity that has been chosen is 'stored' in the designator.
+* To use this, state, instantiate a `VariableDesignator` in your example state machine:
+```python
+target_designator = states.util.designators.core.VariableDesignator(resolve_type=robot_skills.util.entity.Entity)
+```
+* Add the `GenerateRandomTarget` state to your state machine. Use the `target_designator` as an argument. Reconnect the transitions
+* Add a `NavigateToSymbolic` state to your state machine
+```Python
+states.NavigateToSymbolic(robot, entity_designator_area_name_map={target_designator: "in_front_of"},
+                            entity_lookat_designator=target_designator)
+```
+(this will make the robot move to the 'in_front_of' area of the designated furniture object while looking at the same furniture object). Reconnect the transitions once again.
+* Run the example
```

Comment:
Some of the furniture entities don't have a `in_front_of` area, such as `trashbin` and `nightstand`, so when running the example you might get the errors:
```
[ERROR] [1537620868.227588]: Entity 'nightstand': does not have an 'areas' property and the area name is not near'; it it 'in_front_of'.
[ERROR] [1537620868.229535]: Could not get plan from global planner via service call, is the global planner running?
[ERROR] [1537620868.229873]: 'NoneType' object has no attribute 'frame'
```